src/main/java/com/epam/aidial/core/controller/ResourceController.java

```diff
@@ -117,30 +122,39 @@ private Future<?> getResource(ResourceDescription descriptor, boolean hasWriteAc
             return context.respond(HttpStatus.BAD_REQUEST, "Folder not allowed: " + descriptor.getUrl());
         }
 
-        vertx.executeBlocking(() -> service.getResourceWithMetadata(descriptor), false)
-                .onSuccess(pair -> {
-                    if (pair == null) {
-                        context.respond(HttpStatus.NOT_FOUND, "Not found: " + descriptor.getUrl());
-                    } else {
-                        ResourceItemMetadata itemMetadata = pair.getKey();
-                        String body = pair.getValue();
-                        context.putHeader(HttpHeaders.ETAG, itemMetadata.getEtag())
-                                .exposeHeaders();
-
-                        // if resource type is application and caller has no write access - return application data
-                        if (descriptor.getType() == ResourceType.APPLICATION && !hasWriteAccess) {
-                            Application application = ProxyUtil.convertToObject(body, Application.class, true);
-                            ApplicationData applicationData = ApplicationUtil.mapApplication(application);
-                            context.respond(HttpStatus.OK, ProxyUtil.convertToString(applicationData));
-                        } else {
-                            context.respond(HttpStatus.OK, body);
-                        }
-                    }
+        Future<Pair<ResourceItemMetadata, String>> responseFuture;
+
+        if (descriptor.getType() == ResourceType.APPLICATION) {
```

Comment:
It's better to move the block to a single method.

Reply:
Please collaborate. I prefer not having complex inner ifs.

Reply:
```
private Future<Pair<ResourceItemMetadata, String>> getApplication(ResourceDescription descriptor, boolean hasWriteAccess) {
        return vertx.executeBlocking(() -> {
            Pair<ResourceItemMetadata, Application> result = applicationService.getApplication(descriptor);
            ResourceItemMetadata meta = result.getKey();

            Application application = result.getValue();
            String body = hasWriteAccess
                    ? ProxyUtil.convertToString(application, true)
                    : ProxyUtil.convertToString(ApplicationUtil.mapApplication(application));

            return Pair.of(meta, body);

        }, false);
    }
```

makes sense?